Mathlib/NumberTheory/FactorisationProperties.lean

```diff
@@ -0,0 +1,235 @@
+
+/-- `n : ℕ` is a weird number if and only if it is both abundant and pseudoperfect -/
+def WeirdNumber (n : ℕ) : Prop := Abundant n ∧ ¬ Pseudoperfect n
+
+theorem not_pseudoperfect_iff_forall :
+    ¬ Pseudoperfect n ↔ n = 0 ∨ ∀ (s : Finset ℕ), s ∈ powerset (properDivisors n)
+      → ∑ i in s, i ≠ n := by
+    have hn : (¬ 0 < n) ↔ n = 0 := by simp only [not_lt, nonpos_iff_eq_zero]
+    dsimp only [Pseudoperfect]
+    rw [not_and_or, not_exists, hn]
+    constructor
+    · rintro (h1 | h2)
+      · tauto
+      · right
+        intro s
+        have hs : ¬(s ∈ powerset (properDivisors n) ∧ ∑ i in s, i = n) := h2 s
+        rw [not_and] at hs
+        exact hs
+    · rintro (h1 | h2)
+      · tauto
+      · right
+        simp_rw [not_and]
+        exact h2
+
+theorem one_deficient : Deficient 1 := le.refl
+
+theorem two_deficient : Deficient 2 := le.refl
+
+theorem three_deficient : Deficient 3 := by
+  dsimp only [Deficient]
+  norm_num
+
+theorem six_perfect : Perfect 6 := by
+  dsimp only [Perfect]
+  rw [show properDivisors 6 = {1, 2, 3} by rfl]
+  norm_num
+
+theorem twelve_abundant : Abundant 12 := by
+  dsimp only [Abundant]
+  rw [show properDivisors 12 = {1,2,3,4,6} by rfl]
+  norm_num
+
+set_option maxRecDepth 2000 in
+theorem seventy_weird : WeirdNumber 70 := by
+  dsimp only [WeirdNumber, Abundant]
+  rw [not_pseudoperfect_iff_forall]
+  have div70 : properDivisors 70 = {1, 2, 5, 7, 10, 14, 35} := by rfl
+  constructor
+  · rw [div70]
+    repeat norm_num
+  · rw [div70]
+    right
+    intro s hs
+    fin_cases hs <;> decide
+
+lemma deficient_not_abundant_or_perfect (hn : 0 < n) :
+    Deficient n ↔ ¬ Abundant n ∧ ¬ Perfect n := by
+      dsimp only [Perfect, Abundant, Deficient]
+      omega
+
+lemma perfect_not_abundant_or_deficient (hn : 0 < n) :
+    Perfect n ↔ ¬ Abundant n ∧ ¬ Deficient n := by
+      dsimp only [Perfect, Abundant, Deficient]
+      omega
+
+lemma abundant_not_perfect_or_deficient (hn : 0 < n) :
+    Abundant n ↔ ¬ Perfect n ∧ ¬ Deficient n := by
+      dsimp only [Perfect, Abundant, Deficient]
+      omega
+
+/--A positive natural number is either deficient, perfect, or abundant-/
+theorem deficient_or_perfect_or_abundant (hn : 0 < n) :
+    Deficient n ∨ Abundant n ∨ Perfect n := by
+      dsimp only [Perfect, Abundant, Deficient]
+      omega
+
+theorem perfect_pseudoperfect (h : Perfect n) : Pseudoperfect n := by
+  rcases h with ⟨h1, h2⟩
+  exact ⟨h2, properDivisors n, mem_powerset_self (properDivisors n), h1⟩
+
+theorem prime_not_abundant (h : Prime n) : ¬ Abundant n := by
+  intro h1
+  have h2 := sum_properDivisors_eq_one_iff_prime.mpr h
+  have h3 := Nat.lt_trans (Prime.one_lt h) h1
+  omega
+
+theorem prime_not_weird (h : Prime n) : ¬ WeirdNumber n := by
+  simp only [Nat.WeirdNumber, not_and_or]
+  left
+  exact prime_not_abundant n h
+
+theorem not_pseudoperfect_not_perfect (h : ¬ Pseudoperfect n) : ¬ Perfect n := by
+  revert h
+  rw [not_imp_not]
+  exact perfect_pseudoperfect n
```

Comment:
Is this theorem really needed? If we have a theorem that X implies Y, I don't think we really need to spell out that that not-Y implies not-X as well.

Reply:
Okay I'll get rid of this one